---
title: "RBN exploration"
author: Carlos Blanco
output:
  html_document:
    toc: true
    number_sections: true
    toc_float: true
editor_options:
  markdown:
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(purrr)
library(stringr)
library(data.table)
library(EnsDb.Mmusculus.v79)
library(EDASeq)
library(org.Mm.eg.db)
library(clusterProfiler)
library(enrichplot)
library(RColorConesa)
```



First, obtain a per gene classification file for each sample. In these files, each row will be a gene, and the columns will be the counts of reads for each structural category, the total of the read counts, the median of the lengths of each transcript model associated with that gene, and the median number of exons of each transcript model associated with that gene.

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore_not_subsampled", pattern = "B", full.names = TRUE)

process_file <- function(file) {
  data <- read.delim(file, stringsAsFactors = FALSE)
  total_counts <- sum(data$FL)
  data$CPM <- (data$FL / total_counts) * 1e6
  data_filtered <- data %>%
  filter(
    !(grepl("SIRV", chrom)) & 
    !(filter_result == "Artifact") & 
    !(structural_category %in% c("intergenic", "antisense", "fusion", "genic_intron")))
 
  inter_df <- data_filtered %>%
	select(isoform, chrom, associated_gene, structural_category, FL, length, exons, CPM) %>%
	arrange(associated_gene)  
  
  final_df <- inter_df %>%
	group_by(associated_gene, chrom) %>%
	summarise(median_length = median(length),
          	median_exons = median(exons),
          	full_splice_match_counts = sum(FL[structural_category == "full-splice_match"]),
          	incomplete_splice_match_counts = sum(FL[structural_category == "incomplete-splice_match"]),
          	novel_in_catalog_counts = sum(FL[structural_category == "novel_in_catalog"]),
          	novel_not_in_catalog_counts = sum(FL[structural_category == "novel_not_in_catalog"]),
          	genic_counts = sum(FL[structural_category == "genic"]),
          	total_counts = sum(FL),
	          
	          full_splice_match_cpm = sum(CPM[structural_category == "full-splice_match"]),
          	incomplete_splice_match_cpm = sum(CPM[structural_category == "incomplete-splice_match"]),
          	novel_in_catalog_cpm = sum(CPM[structural_category == "novel_in_catalog"]),
          	novel_not_in_catalog_cpm = sum(CPM[structural_category == "novel_not_in_catalog"]),
          	genic_cpm = sum(CPM[structural_category == "genic"]),
          	total_cpm = sum(CPM))
 
  sample_name <- sub("_.*", "", basename(file))
  output_name <- paste0(sample_name, "_per_gene_filtered_classification.txt")
  write.table(final_df, file = paste0("/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications_nanopore_not_subsampled/", output_name), sep = "\t", row.names = FALSE)
}

dir <- "/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications_nanopore_not_subsampled"
if (!dir.exists(dir)) {
  dir.create(dir)
}

lapply(files, process_file)

my_theme <- theme(
    axis.text.x = element_text(size = 20),
    axis.text.y = element_text(size = 20),
    axis.title.x = element_text(size = 25),
    axis.title.y = element_text(size = 25),
    legend.title = element_text(size = 20),
    legend.text = element_text(size = 18),
    legend.key.size = unit(0.5, 'cm'),
    panel.background = element_rect(fill = "transparent", colour = NA), 
    plot.background = element_rect(fill = "transparent", colour = NA),
    legend.background = element_rect(fill = "transparent", colour = NA),
    plot.margin = margin(t = 10, r = 10, b = 10, l = 10),
    )
```



Once we have the per gene classifications with the counts, median length and median number of exons, we will calculate the noisy reads ratio difference per gene, alongside its length and number of exons. For this we first filter the genes that have at least 10 reads assigned as FSM in every sample, and also appear in every sample (to avoid genes that are underrepresented) Then, we obtain the ratio for each structural category of each gene, then the mean of the ratio of each structural category for the samples in the young group of mice and the samples in the old group of mice, independently. Then, the noisy reads ratio (NIC+NNC/NIC+NNC+FSM) is calculated for each gene of each age group and the difference (old - young) is obtained.

For the length and number of exons of the gene, we get the median length and median number of exons of each gene for each sample and compute the average for each age group. Then, from that average for each age group, a second average is calculated to obtain the final length and number of exons per gene.

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications_nanopore_not_subsampled", full.names = TRUE)

files <- files[!grepl("^/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications_nanopore_not_subsampled/B152", files)]

filtered_data <- list()

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
  data <- data[data$full_splice_match_counts >= 10,]
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-5]")]
old_files <- files[str_detect(basename(files), "^B15[1-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  
          	.groups = "drop") %>%
	arrange(associated_gene)  
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.005, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count") +
    scale_fill_conesa(name = "Greater in old", palette = "complete", labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 300, label = paste(porcentaje_below_0, "%"), color = "#EE446F", size = 10) +
    annotate("text", x = 0.1, y = 300, label = paste(porcentaje_above_0, "%"), color = "#15918A", size = 10) +
    theme_minimal() +
    my_theme
  return(histograma)
}
```


```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
ggsave("nanopore_all_data_RBNdiff.png", histogram, path = "/home/carblan/Documents/R_files/TFM/figures/RBNdiff_histograms/ONT", width = 10, height = 7)
```



To understand the relationship between gene length or number of exons with RBN (RNA biological noise)

```{r}
per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

length_correlation_spearman <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_length, method = "spearman")
length_correlation_pearson <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_length, method = "pearson")

ggplot(per_gene_RBN, aes(x = noisy_reads_ratio_diff, y = average_length)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = "Correlation between per gene average transcript model length and Noisy Reads Ratio Difference (all samples)",
       x = "Noisy Reads Ratio Difference",
       y = "Weighted Length Difference") +
  annotate("text", x = min(per_gene_RBN$noisy_reads_ratio_diff), y = max(per_gene_RBN$average_length), 
           label = paste("Spearman: ", round(length_correlation_spearman, 3), "\nPearson: ", round(length_correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))


exons_correlation_spearman <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_exons, method = "spearman")
exons_correlation_pearson <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_exons, method = "pearson")

ggplot(per_gene_RBN, aes(x = noisy_reads_ratio_diff, y = average_exons)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = "Correlation between per gene average transcript models exons number and Noisy Reads Ratio Difference (all samples)",
       x = "Noisy Reads Ratio Difference",
       y = "Weighted Length Difference") +
  annotate("text", x = min(per_gene_RBN$noisy_reads_ratio_diff), y = max(per_gene_RBN$average_exons), 
           label = paste("Spearman: ", round(exons_correlation_spearman, 3), "\nPearson: ", round(exons_correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))

```



The same is done, taking the length and the gc content of the gene from a reference, instead of calculating the length and the number of exons from the transcript models data.

```{r}
per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

symbols <- per_gene_RBN$associated_gene

ids <- AnnotationDbi::select(EnsDb.Mmusculus.v79,  
   	keys = symbols,
   	columns = c("ENTREZID", "SYMBOL", "GENEID"),
   	keytype = "SYMBOL")

lengths_gc <- getGeneLengthAndGCContent(ids$GENEID, "mmu", mode="org.db") #Choose mart 4 (mus musculus) if you select mode=biomart (not working because lack of memory?)

merged_df <- merge(per_gene_RBN, ids, by.x = "associated_gene", by.y = "SYMBOL")
merged_df <- merge(merged_df, lengths_gc, by.x = "GENEID", by.y = "row.names")

final_df <- merged_df[, c("associated_gene", "chrom", "length", "gc", "noisy_reads_ratio_diff")][order(merged_df$associated_gene),]

final_df <- final_df %>%
  dplyr::filter(!((final_df$length == "NA")))

length_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$length, method = "spearman")
print(length_correlation)

gc_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$gc, method = "spearman")
print(gc_correlation)
```



```{r}
```



This is also important just to check if the noisy genes are noisy because they
are truly noisy or because they just have low read count (even if the genes are filtered by
FSM >= 10, the noisy genes could just be noisy because the overall read count of the gene is low)

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled", pattern=".txt", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

median_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, FUN = median)

names(median_cpm_data)[names(median_cpm_data) == "cpm"] <- "median_cpm"


per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

combined_data <- merge(median_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "spearman")
cat("Median cpm counts and noise difference correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "pearson")
cat("Median cpm counts and noise difference correlation (Pearson):\n", correlation, "\n")
```


```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled", pattern=".txt", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

total_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, sum)

names(total_cpm_data)[names(total_cpm_data) == "cpm"] <- "total_cpm"


per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

combined_data <- merge(total_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$total_cpm, method = "spearman")
cat("Total cpm counts and noise difference correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$total_cpm, method = "pearson")
cat("Gene counts and noise difference correlation (Pearson):\n", correlation, "\n")
```



Now we do the same but only for pool 1 samples, as we know that the percentage of genes that show more noise in old mice than young mice is 60% in pool 1 but 40% in pool 2. The genes filtered are not exclusive to pool 1, these are only all of the genes shown by all samples from pool 1 that also have 10 or more counts assigned as FSM.

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications_nanopore_not_subsampled",
                	full.names = TRUE)
pool_1_files <- files[str_detect(basename(files), "^B3[1-3]|^B15[1-2]")]
filtered_data <- list()

for (file in pool_1_files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in pool_1_files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-3]")]
old_files <- files[str_detect(basename(files), "^B15[1-2]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.2, y = 1500, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.2, y = 1500, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
```



To understand the relationship between gene length or number of exons with RBN (RNA biological noise)

```{r}
per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

length_correlation <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_length, method = "spearman")
print(length_correlation)

exons_correlation <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_exons, method = "spearman")
print(exons_correlation)

```


The same is done, taking the length and the gc content of the gene from a reference, instead of calculating the length and the number of exons from the transcript models data.

```{r}

per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

symbols <- per_gene_RBN$associated_gene

ids <- AnnotationDbi::select(EnsDb.Mmusculus.v79,  
   	keys = symbols,
   	columns = c("ENTREZID", "SYMBOL", "GENEID"),
   	keytype = "SYMBOL")

lengths_gc <- getGeneLengthAndGCContent(ids$GENEID, "mmu", mode="org.db") #Choose mart 4 (mus musculus) if you select mode=biomart (not working because lack of memory?)

merged_df <- merge(per_gene_RBN, ids, by.x = "associated_gene", by.y = "SYMBOL")
merged_df <- merge(merged_df, lengths_gc, by.x = "GENEID", by.y = "row.names")

final_df <- merged_df[, c("associated_gene", "chrom", "length", "gc", "noisy_reads_ratio_diff")][order(merged_df$associated_gene),]

final_df <- final_df %>%
  dplyr::filter(!((final_df$length == "NA")))

length_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$length, method = "spearman")
print(length_correlation)

gc_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$gc, method = "spearman")
print(gc_correlation)
```



This is also important just to check if the noisy genes are noisy because they
are truly noisy or because they just have low read count (even if the genes are filtered by
FSM >= 10, the noisy genes could just be noisy because the overall read count of the gene is low)

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore", pattern=".txt", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

median_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, FUN = median)

names(median_cpm_data)[names(median_cpm_data) == "cpm"] <- "median_cpm"


per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

combined_data <- merge(median_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "spearman")
cat("Gene counts and length correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "pearson")
cat("Gene counts and length correlation (Pearson):\n", correlation, "\n")
```



Now we do the same but only for pool 2 samples, as we know that the percentage of genes that show more noise in old mice than young mice is 60% in pool 1 but 40% in pool 2. The genes filtered are not exclusive to pool 2, these are only all of the genes shown by all samples from pool 2 that also have 10 or more counts assigned as FSM.

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications_nanopore_not_subsampled",
                	full.names = TRUE)
pool_2_files <- files[str_detect(basename(files), "^B3[4-5]|^B15[3-4]")]
filtered_data <- list()

for (file in pool_2_files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in pool_2_files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[4-5]")]
old_files <- files[str_detect(basename(files), "^B15[3-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar por associated_gene
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.25, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.25, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
```



To understand the relationship between gene length or number of exons with RBN difference (RNA biological noise difference between old and young samples)

```{r}
per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

length_correlation <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_length, method = "spearman")
print(length_correlation)

exons_correlation <- cor(per_gene_RBN$noisy_reads_ratio_diff, per_gene_RBN$average_exons, method = "spearman")
print(exons_correlation)

```


The same is done, taking the length and the gc content of the gene from a reference, instead of calculating the length and the number of exons from the transcript models data.

```{r}

per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

symbols <- per_gene_RBN$associated_gene

ids <- AnnotationDbi::select(EnsDb.Mmusculus.v79,  
   	keys = symbols,
   	columns = c("ENTREZID", "SYMBOL", "GENEID"),
   	keytype = "SYMBOL")

lengths_gc <- getGeneLengthAndGCContent(ids$GENEID, "mmu", mode="org.db") #Choose mart 4 (mus musculus) if you select mode=biomart (not working because lack of memory?)

merged_df <- merge(per_gene_RBN, ids, by.x = "associated_gene", by.y = "SYMBOL")
merged_df <- merge(merged_df, lengths_gc, by.x = "GENEID", by.y = "row.names")

final_df <- merged_df[, c("associated_gene", "chrom", "length", "gc", "noisy_reads_ratio_diff")][order(merged_df$associated_gene),]

final_df <- final_df %>%
  dplyr::filter(!((final_df$length == "NA")))

length_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$length, method = "spearman")
print(length_correlation)

gc_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$gc, method = "spearman")
print(gc_correlation)
```



This is also important just to check if the noisy genes are noisy because they
are truly noisy or because they just have low read count (even if the genes are filtered by
FSM >= 10, the noisy genes could just be noisy because the overall read count of the gene is low)

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes_nanopore", pattern=".txt", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

median_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, FUN = median)

names(median_cpm_data)[names(median_cpm_data) == "cpm"] <- "median_cpm"


per_gene_RBN <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)

combined_data <- merge(median_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "spearman")
cat("Gene counts and length correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "pearson")
cat("Gene counts and length correlation (Pearson):\n", correlation, "\n")
```



Now we will do the same than with the intersection, but taking the counts only from pool 1. With this we want to check whether this genes behave the same way with the pool 1 data or behave the same as in the intersection. The percentages should be the same with the same genes, cause technically the pool 1 has selected longer transcripts, but the genes are the same, so the transcripts selected in either pool should have the same ratios. 

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool1 <- files[str_detect(basename(files), "^B3[1-3]")]
old_files_pool1 <- files[str_detect(basename(files), "^B15[1-2]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios_pool1 <- calculate_mean_ratios(young_files_pool1)

old_mean_ratios_pool1 <- calculate_mean_ratios(old_files_pool1)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(young_mean_ratios_pool1)
old_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(old_mean_ratios_pool1)


calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_pool1 <- calculate_difference(young_noisy_reads_ratios_pool1, old_noisy_reads_ratios_pool1)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count") +
    scale_fill_conesa(name = "Greater in old", palette = "complete", labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.2, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#EE446F", size = 10) +
    annotate("text", x = 0.2, y = 1250, label = paste(porcentaje_above_0, "%"), color = "#15918A", size = 10) +
    my_theme
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(noisy_reads_ratio_difference_pool1)
print(histogram)
ggsave("pacbio_pool1_RBNdiff.png", histogram, path = "/home/carblan/Documents/R_files/TFM/figures/RBNdiff_histograms", width = 10, height = 7)
```



To understand the relationship between gene length or number of exons with RBN (RNA biological noise)

```{r}
length_correlation <- cor(noisy_reads_ratio_difference_pool1$noisy_reads_ratio_diff, noisy_reads_ratio_difference_pool1$average_length, method = "spearman")
print(length_correlation)

exons_correlation <- cor(noisy_reads_ratio_difference_pool1$noisy_reads_ratio_diff, noisy_reads_ratio_difference_pool1$average_exons, method = "spearman")
print(exons_correlation)

```


The same is done, taking the length and the gc content of the gene from a reference, instead of calculating the length and the number of exons from the transcript models data.

```{r}
symbols <- noisy_reads_ratio_difference_pool1$associated_gene

ids <- AnnotationDbi::select(EnsDb.Mmusculus.v79,  
   	keys = symbols,
   	columns = c("ENTREZID", "SYMBOL", "GENEID"),
   	keytype = "SYMBOL")

lengths_gc <- getGeneLengthAndGCContent(ids$GENEID, "mmu", mode="org.db") #Choose mart 4 (mus musculus) if you select mode=biomart (not working because lack of memory?)

merged_df <- merge(noisy_reads_ratio_difference_pool1, ids, by.x = "associated_gene", by.y = "SYMBOL")
merged_df <- merge(merged_df, lengths_gc, by.x = "GENEID", by.y = "row.names")

final_df <- merged_df[, c("associated_gene", "chrom", "length", "gc", "noisy_reads_ratio_diff")][order(merged_df$associated_gene),]

final_df <- final_df %>%
  dplyr::filter(!((final_df$length == "NA")))

length_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$length, method = "spearman")
print(length_correlation)

gc_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$gc, method = "spearman")
print(gc_correlation)
```



This is also important just to check if the noisy genes are noisy because they
are truly noisy or because they just have low read count (even if the genes are filtered by
FSM >= 10, the noisy genes could just be noisy because the overall read count of the gene is low)

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled", pattern="B3[1-3]|B15[1-2]", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

median_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, FUN = median)

names(median_cpm_data)[names(median_cpm_data) == "cpm"] <- "median_cpm"


per_gene_RBN <- noisy_reads_ratio_difference_pool1

combined_data <- merge(median_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "spearman")
cat("Median cpm counts and noise difference correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "pearson")
cat("Median cpm counts and noise difference correlation (Pearson):\n", correlation, "\n")
```


```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled", pattern="B3[1-3]|B15[1-2]", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

total_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, sum)

names(total_cpm_data)[names(total_cpm_data) == "cpm"] <- "total_cpm"


per_gene_RBN <- noisy_reads_ratio_difference_pool1

combined_data <- merge(total_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$total_cpm, method = "spearman")
cat("Total cpm counts and noise difference correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$total_cpm, method = "pearson")
cat("Total cpm counts and noise difference correlation (Pearson):\n", correlation, "\n")
```



```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49]
  
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  
  df$sample <- sub("_.*", "", basename(file))
  
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore_not_subsampled", pattern = "B3[1-3]|B15[1-2]", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$age_group <- ifelse(combined_df$sample %in% c("B151", "B152"), "old", "young")

combined_df_young <- combined_df %>% filter(sample %in% c("B31", "B32", "B33"))
combined_df_old <- combined_df %>% filter(sample %in% c("B151", "B152"))


weighted_length_df_young <- combined_df_young %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(age_group = "young")

weighted_length_df_old <- combined_df_old %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(age_group = "old")

weighted_length_df <- merge(noisy_reads_ratio_difference_pool1, weighted_length_df_old, by = "associated_gene")


top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "spearman")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "pearson")


ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "pool 1 samples, both age groups transcripts for weigthed length calculation"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))

top_10_percent_genes <- weighted_length_df %>%
  filter(top_10_percent == "Top 10%")

correlation_spearman <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms pool 1 samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms pool 1 samples) and noise difference correlation (Spearman):\n", correlation_pearson, "\n")

ggplot(top_10_percent_genes, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "pool 1 samples, top 10% genes with hight diff"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(top_10_percent_genes$noisy_reads_ratio_diff), y = max(top_10_percent_genes$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))

```


Now we will do the same than with the intersection, but taking the counts only from pool 2. With this we want to check whether this genes behave the same way with the pool 2 data or behave the same as in the intersection. This way we can confirm that the noise compensation effect we can see with the genes in the intersection is real or if its not really a compensatory effect. If we see the same 40/60 percentage of noisy genes then we can confirm that the noise of one pool is compensating the noise of the other, resulting in the 50/50 percentage of noisy genes. 

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool2 <- files[str_detect(basename(files), "^B3[4-5]")]
old_files_pool2 <- files[str_detect(basename(files), "^B15[3-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios_pool2 <- calculate_mean_ratios(young_files_pool2)

old_mean_ratios_pool2 <- calculate_mean_ratios(old_files_pool2)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(young_mean_ratios_pool2)
old_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(old_mean_ratios_pool2)


calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_pool2 <- calculate_difference(young_noisy_reads_ratios_pool2, old_noisy_reads_ratios_pool2)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.025, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "Genes in the intersection of both pools (but only pool 2 read counts)")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.15, y = 750, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 12) +
    annotate("text", x = 0.15, y = 750, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 12) +
    theme(
      axis.text.x = element_text(size = 17),
      axis.text.y = element_text(size = 17),
      axis.title.x = element_text(size = 20),
      axis.title.y = element_text(size = 20),
      plot.title = element_text(size = 25, hjust = 0.5),
      legend.title = element_text(size = 17),
      legend.text = element_text(size = 15))

  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(noisy_reads_ratio_difference_pool2)
print(histogram)
```



To understand the relationship between gene length or number of exons with RBN (RNA biological noise)

```{r}
length_correlation <- cor(noisy_reads_ratio_difference_pool2$noisy_reads_ratio_diff, noisy_reads_ratio_difference_pool2$average_length, method = "spearman")
print(length_correlation)

exons_correlation <- cor(noisy_reads_ratio_difference_pool2$noisy_reads_ratio_diff, noisy_reads_ratio_difference_pool2$average_exons, method = "spearman")
print(exons_correlation)

```


The same is done, taking the length and the gc content of the gene from a reference, instead of calculating the length and the number of exons from the transcript models data.

```{r}
symbols <- noisy_reads_ratio_difference_pool2$associated_gene

ids <- AnnotationDbi::select(EnsDb.Mmusculus.v79,  
   	keys = symbols,
   	columns = c("ENTREZID", "SYMBOL", "GENEID"),
   	keytype = "SYMBOL")

lengths_gc <- getGeneLengthAndGCContent(ids$GENEID, "mmu", mode="org.db") #Choose mart 4 (mus musculus) if you select mode=biomart (not working because lack of memory?)

merged_df <- merge(noisy_reads_ratio_difference_pool2, ids, by.x = "associated_gene", by.y = "SYMBOL")
merged_df <- merge(merged_df, lengths_gc, by.x = "GENEID", by.y = "row.names")

final_df <- merged_df[, c("associated_gene", "chrom", "length", "gc", "noisy_reads_ratio_diff")][order(merged_df$associated_gene),]

final_df <- final_df %>%
  dplyr::filter(!((final_df$length == "NA")))

length_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$length, method = "spearman")
print(length_correlation)

gc_correlation <- cor(final_df$noisy_reads_ratio_diff, final_df$gc, method = "spearman")
print(gc_correlation)
```


This is also important just to check if the noisy genes are noisy because they
are truly noisy or because they just have low read count (even if the genes are filtered by
FSM >= 10, the noisy genes could just be noisy because the overall read count of the gene is low)

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled", pattern="B3[4-5]|B15[3-4]", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

median_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, FUN = median)

names(median_cpm_data)[names(median_cpm_data) == "cpm"] <- "median_cpm"


per_gene_RBN <- noisy_reads_ratio_difference_pool2

combined_data <- merge(median_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "spearman")
cat("Median cpm counts and noise difference correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$median_cpm, method = "pearson")
cat("Median cpm counts and noise difference correlation (Pearson):\n", correlation, "\n")
```




```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled", pattern="B3[4-5]|B15[3-4]", full.names = TRUE)

normalize_counts <- function(file) {
  data <- read.table(file, header = TRUE)
  data$cpm <- (data$total_counts / sum(data$total_counts)) * 1e6
  return(data)
}

all_data <- lapply(files, function(file) {
  data <- normalize_counts(file)
  return(data[, c("associated_gene", "chrom", "cpm")])
})

combined_data <- do.call(rbind, all_data)

total_cpm_data <- aggregate(cpm ~ associated_gene + chrom, data = combined_data, sum)

names(total_cpm_data)[names(total_cpm_data) == "cpm"] <- "total_cpm"


per_gene_RBN <- noisy_reads_ratio_difference_pool2

combined_data <- merge(total_cpm_data, per_gene_RBN, by = c("associated_gene", "chrom"))

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$total_cpm, method = "spearman")
cat("Total cpm counts and noise difference correlation (Spearman):\n", correlation, "\n")

correlation <- cor(combined_data$noisy_reads_ratio_diff, combined_data$total_cpm, method = "pearson")
cat("Total cpm counts and noise difference correlation (Pearson):\n", correlation, "\n")
```



```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49]
  
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  
  df$sample <- sub("_.*", "", basename(file))
  
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", 
                    pattern = "B3[4-5]|B15[3-4]", full.names = TRUE)

filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$age_group <- ifelse(combined_df$sample %in% c("B153", "B154"), "old", "young")

combined_df_young <- combined_df %>% filter(sample %in% c("B34", "B35"))
combined_df_old <- combined_df %>% filter(sample %in% c("B153", "B154"))


weighted_length_df_young <- combined_df_young %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(age_group = "young")

weighted_length_df_old <- combined_df_old %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(age_group = "old")

weighted_length_df <- merge(noisy_reads_ratio_difference_pool2, weighted_length_df_old, by = "associated_gene")


top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "spearman")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "pearson")


ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "pool 2 samples, both age groups transcripts for weigthed length calculation"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))

top_10_percent_genes <- weighted_length_df %>%
  filter(top_10_percent == "Top 10%")

correlation_spearman <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "spearman")

correlation_pearson <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "pearson")

ggplot(top_10_percent_genes, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "pool 2 samples, top 10% genes with hight diff"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(top_10_percent_genes$noisy_reads_ratio_diff), y = max(top_10_percent_genes$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))
```



```{r, warning=FALSE}
merged_transcriptome_clas <- read.table("/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore/brain_merged_transcriptome_classification.txt", head = TRUE)
merged_transcriptome_counts <- read.table("/home/carblan/Documents/R_files/TFM/fl_counts_nanopore/merged_filtered_samples.counts.tsv", head = TRUE)

merged_transcriptome_clas <- merged_transcriptome_clas[, c("isoform", "chrom", "length", "exons", "associated_gene", "structural_category")]

df_combined <- merge(merged_transcriptome_clas, merged_transcriptome_counts, by.x = "isoform", by.y = "transcript_id") # Cambia "transcript_id" por el nombre de la columna de IDs en df_counts si es diferente

df_combined <- df_combined %>%
  gather(key = "sample", value = "count", -isoform, -chrom, -length, -exons, -associated_gene, -structural_category) %>%
  group_by(associated_gene, structural_category, sample) %>%
  summarise(total_count = sum(count, na.rm = TRUE)) %>%
  spread(key = "sample", value = "total_count")

df_long <- df_combined %>%
  gather(key = "sample", value = "count", -associated_gene, -structural_category)

df_full_splice <- df_long %>%
  filter(structural_category == "full-splice_match")

genes_to_keep <- df_full_splice %>%
  group_by(associated_gene) %>%
  summarise(min_count = min(count, na.rm = TRUE)) %>%
  filter(min_count >= 10) %>%
  pull(associated_gene)

df_filtered <- df_summary %>%
  filter(associated_gene %in% genes_to_keep)

df_long_filtered <- df_filtered %>%
  gather(key = "sample", value = "count", -associated_gene, -structural_category)

df_normalized <- df_long_filtered %>%
  group_by(associated_gene, sample) %>%
  mutate(total_count = sum(count, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(normalized_count = count / total_count) %>%
  select(-total_count)

group_old <- c("B151.filtered", "B152.filtered", "B153.filtered", "B154.filtered")
group_young <- c("B31.filtered", "B32.filtered", "B33.filtered", "B34.filtered", "B35.filtered")

df_normalized <- df_normalized %>%
  mutate(age_group = ifelse(sample %in% group_old, "old", "young"))

df_means <- df_normalized %>%
  group_by(associated_gene, structural_category, age_group) %>%
  summarise(mean_normalized_count = mean(normalized_count, na.rm = TRUE)) %>%
  ungroup()

df_means_wide <- df_means %>%
  spread(key = "age_group", value = "mean_normalized_count")

calculate_noisy_reads_ratio <- function(full_splice, novel_in_catalog, novel_not_in_catalog) {
  (novel_in_catalog + novel_not_in_catalog) / (full_splice + novel_in_catalog + novel_not_in_catalog)
}

df_old <- df_means_wide %>%
  filter(structural_category %in% c("full-splice_match", "novel_in_catalog", "novel_not_in_catalog")) %>%
  select(associated_gene, structural_category, old)

df_young <- df_means_wide %>%
  filter(structural_category %in% c("full-splice_match", "novel_in_catalog", "novel_not_in_catalog")) %>%
  select(associated_gene, structural_category, young)

df_old_wide <- df_old %>%
  spread(key = structural_category, value = old, fill = 0)

df_young_wide <- df_young %>%
  spread(key = structural_category, value = young, fill = 0)

df_old_wide <- df_old_wide %>%
  mutate(
    `full-splice_match` = ifelse(is.na(`full-splice_match`), 0, `full-splice_match`),
    novel_in_catalog = ifelse(is.na(novel_in_catalog), 0, novel_in_catalog),
    novel_not_in_catalog = ifelse(is.na(novel_not_in_catalog), 0, novel_not_in_catalog)
  )

df_young_wide <- df_young_wide %>%
  mutate(
    `full-splice_match` = ifelse(is.na(`full-splice_match`), 0, `full-splice_match`),
    novel_in_catalog = ifelse(is.na(novel_in_catalog), 0, novel_in_catalog),
    novel_not_in_catalog = ifelse(is.na(novel_not_in_catalog), 0, novel_not_in_catalog)
  )

df_old_wide <- df_old_wide %>%
  mutate(
    old_noisy_reads_ratio = calculate_noisy_reads_ratio(
      `full-splice_match`, novel_in_catalog, novel_not_in_catalog
    )
  )

df_young_wide <- df_young_wide %>%
  mutate(
    young_noisy_reads_ratio = calculate_noisy_reads_ratio(
      `full-splice_match`, novel_in_catalog, novel_not_in_catalog
    )
  )

df_noisy_reads_ratio <- merge(
  df_old_wide[, c("associated_gene", "old_noisy_reads_ratio")],
  df_young_wide[, c("associated_gene", "young_noisy_reads_ratio")],
  by = "associated_gene",
  all = TRUE
) %>%
  mutate(noisy_reads_ratio_diff = old_noisy_reads_ratio - young_noisy_reads_ratio) %>%
  select(associated_gene, noisy_reads_ratio_diff)
```


```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young (merged transcriptome):\n", "NNC + NIC / NNC + NIC + FSM\n", "Genes in the intersection of both pools")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.1, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(df_noisy_reads_ratio)
print(histogram)
```


To understand the relationship between weighted gene length with RBN (RNA biological noise)

```{r}
merged_transcriptome_clas <- read.table("/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore/brain_merged_transcriptome_classification.txt", head = TRUE)
merged_transcriptome_counts <- read.table("/home/carblan/Documents/R_files/TFM/fl_counts_nanopore/merged_filtered_samples.counts.tsv", head = TRUE)

merged_transcriptome_clas <- merged_transcriptome_clas %>%
  select(isoform, associated_gene, length)

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  gather(key = "sample", value = "count", -transcript_id)
merged_transcriptome_counts <- merged_transcriptome_counts %>%
  rename(isoform = transcript_id)

merged_data <- merge(merged_transcriptome_clas, merged_transcriptome_counts, by = "isoform")

weighted_lengths <- merged_data %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * count) / sum(count))

noisy_reads_ratios <- read.table("/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
final_data <- merge(weighted_lengths, noisy_reads_ratios, by = "associated_gene")

correlation_result <- cor(final_data$weighted_length, final_data$noisy_reads_ratio_diff, method = "spearman")
print(correlation_result)
```




```{r, warning=FALSE}
df_normalized <- df_long_filtered %>%
  group_by(associated_gene, sample) %>%
  mutate(total_count = sum(count, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(normalized_count = count / total_count) %>%
  select(-total_count)

group_old <- c("B151.filtered", "B152.filtered")
group_young <- c("B31.filtered", "B32.filtered", "B33.filtered")

df_normalized <- df_normalized %>%
  mutate(age_group = ifelse(sample %in% group_old, "old", "young"))

df_means <- df_normalized %>%
  group_by(associated_gene, structural_category, age_group) %>%
  summarise(mean_normalized_count = mean(normalized_count, na.rm = TRUE)) %>%
  ungroup()

df_means_wide <- df_means %>%
  spread(key = "age_group", value = "mean_normalized_count")

calculate_noisy_reads_ratio <- function(full_splice, novel_in_catalog, novel_not_in_catalog) {
  (novel_in_catalog + novel_not_in_catalog) / (full_splice + novel_in_catalog + novel_not_in_catalog)
}

df_old <- df_means_wide %>%
  filter(structural_category %in% c("full-splice_match", "novel_in_catalog", "novel_not_in_catalog")) %>%
  select(associated_gene, structural_category, old)

df_young <- df_means_wide %>%
  filter(structural_category %in% c("full-splice_match", "novel_in_catalog", "novel_not_in_catalog")) %>%
  select(associated_gene, structural_category, young)

df_old_wide <- df_old %>%
  spread(key = structural_category, value = old, fill = 0)

df_young_wide <- df_young %>%
  spread(key = structural_category, value = young, fill = 0)

df_old_wide <- df_old_wide %>%
  mutate(
    `full-splice_match` = ifelse(is.na(`full-splice_match`), 0, `full-splice_match`),
    novel_in_catalog = ifelse(is.na(novel_in_catalog), 0, novel_in_catalog),
    novel_not_in_catalog = ifelse(is.na(novel_not_in_catalog), 0, novel_not_in_catalog)
  )

df_young_wide <- df_young_wide %>%
  mutate(
    `full-splice_match` = ifelse(is.na(`full-splice_match`), 0, `full-splice_match`),
    novel_in_catalog = ifelse(is.na(novel_in_catalog), 0, novel_in_catalog),
    novel_not_in_catalog = ifelse(is.na(novel_not_in_catalog), 0, novel_not_in_catalog)
  )

df_old_wide <- df_old_wide %>%
  mutate(
    old_noisy_reads_ratio = calculate_noisy_reads_ratio(
      `full-splice_match`, novel_in_catalog, novel_not_in_catalog
    )
  )

df_young_wide <- df_young_wide %>%
  mutate(
    young_noisy_reads_ratio = calculate_noisy_reads_ratio(
      `full-splice_match`, novel_in_catalog, novel_not_in_catalog
    )
  )

df_noisy_reads_ratio <- merge(
  df_old_wide[, c("associated_gene", "old_noisy_reads_ratio")],
  df_young_wide[, c("associated_gene", "young_noisy_reads_ratio")],
  by = "associated_gene",
  all = TRUE
) %>%
  mutate(noisy_reads_ratio_diff = old_noisy_reads_ratio - young_noisy_reads_ratio) %>%
  select(associated_gene, noisy_reads_ratio_diff)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young (merged transcriptome):\n", "NNC + NIC / NNC + NIC + FSM\n", "Genes in the intersection of both pools")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.1, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(df_noisy_reads_ratio)
print(histogram)
```


To understand the relationship between weighted gene length with RBN (RNA biological noise)

```{r}
merged_transcriptome_clas <- read.table("/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore/brain_merged_transcriptome_classification.txt", header = TRUE)
merged_transcriptome_counts <- read.table("/home/carblan/Documents/R_files/TFM/fl_counts/merged_filtered_samples.counts.tsv", header = TRUE)

merged_transcriptome_clas <- merged_transcriptome_clas %>%
  select(isoform, associated_gene, length)

selected_samples <- c("B31.filtered", "B32.filtered", "B33.filtered", "B151.filtered", "B152.filtered")

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  gather(key = "sample", value = "count", -transcript_id)

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  filter(sample %in% selected_samples)

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  rename(isoform = transcript_id)

merged_data <- merge(merged_transcriptome_clas, merged_transcriptome_counts, by = "isoform")

weighted_lengths <- merged_data %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * count) / sum(count))

final_data <- merge(weighted_lengths, noisy_reads_ratio_difference_pool1, by = "associated_gene")

correlation_result <- cor(final_data$weighted_length, final_data$noisy_reads_ratio_diff, method = "spearman")
print(correlation_result)
```



```{r, warning=FALSE}
df_normalized <- df_long_filtered %>%
  group_by(associated_gene, sample) %>%
  mutate(total_count = sum(count, na.rm = TRUE)) %>%
  ungroup() %>%
  mutate(normalized_count = count / total_count) %>%
  select(-total_count)

group_old <- c("B153.filtered", "B154.filtered")
group_young <- c("B34.filtered", "B35.filtered")

df_normalized <- df_normalized %>%
  mutate(age_group = ifelse(sample %in% group_old, "old", "young"))

df_means <- df_normalized %>%
  group_by(associated_gene, structural_category, age_group) %>%
  summarise(mean_normalized_count = mean(normalized_count, na.rm = TRUE)) %>%
  ungroup()

df_means_wide <- df_means %>%
  spread(key = "age_group", value = "mean_normalized_count")

calculate_noisy_reads_ratio <- function(full_splice, novel_in_catalog, novel_not_in_catalog) {
  (novel_in_catalog + novel_not_in_catalog) / (full_splice + novel_in_catalog + novel_not_in_catalog)
}

df_old <- df_means_wide %>%
  filter(structural_category %in% c("full-splice_match", "novel_in_catalog", "novel_not_in_catalog")) %>%
  select(associated_gene, structural_category, old)

df_young <- df_means_wide %>%
  filter(structural_category %in% c("full-splice_match", "novel_in_catalog", "novel_not_in_catalog")) %>%
  select(associated_gene, structural_category, young)

df_old_wide <- df_old %>%
  spread(key = structural_category, value = old, fill = 0)

df_young_wide <- df_young %>%
  spread(key = structural_category, value = young, fill = 0)

df_old_wide <- df_old_wide %>%
  mutate(
    `full-splice_match` = ifelse(is.na(`full-splice_match`), 0, `full-splice_match`),
    novel_in_catalog = ifelse(is.na(novel_in_catalog), 0, novel_in_catalog),
    novel_not_in_catalog = ifelse(is.na(novel_not_in_catalog), 0, novel_not_in_catalog)
  )

df_young_wide <- df_young_wide %>%
  mutate(
    `full-splice_match` = ifelse(is.na(`full-splice_match`), 0, `full-splice_match`),
    novel_in_catalog = ifelse(is.na(novel_in_catalog), 0, novel_in_catalog),
    novel_not_in_catalog = ifelse(is.na(novel_not_in_catalog), 0, novel_not_in_catalog)
  )

df_old_wide <- df_old_wide %>%
  mutate(
    old_noisy_reads_ratio = calculate_noisy_reads_ratio(
      `full-splice_match`, novel_in_catalog, novel_not_in_catalog
    )
  )

df_young_wide <- df_young_wide %>%
  mutate(
    young_noisy_reads_ratio = calculate_noisy_reads_ratio(
      `full-splice_match`, novel_in_catalog, novel_not_in_catalog
    )
  )

df_noisy_reads_ratio <- merge(
  df_old_wide[, c("associated_gene", "old_noisy_reads_ratio")],
  df_young_wide[, c("associated_gene", "young_noisy_reads_ratio")],
  by = "associated_gene",
  all = TRUE
) %>%
  mutate(noisy_reads_ratio_diff = old_noisy_reads_ratio - young_noisy_reads_ratio) %>%
  select(associated_gene, noisy_reads_ratio_diff)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young (merged transcriptome):\n", "NNC + NIC / NNC + NIC + FSM\n", "Genes in the intersection of both pools")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.1, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(df_noisy_reads_ratio)
print(histogram)
```



To understand the relationship between weighted gene length with RBN (RNA biological noise)

```{r}
merged_transcriptome_clas <- read.table("/home/carblan/Documents/R_files/TFM/filtered_classifications/brain_merged_transcriptome_classification.txt", header = TRUE)
merged_transcriptome_counts <- read.table("/home/carblan/Documents/R_files/TFM/fl_counts/merged_filtered_samples.counts.tsv", header = TRUE)

merged_transcriptome_clas <- merged_transcriptome_clas %>%
  select(isoform, associated_gene, length)

selected_samples <- c("B34.filtered", "B35.filtered", "B153.filtered", "B154.filtered")

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  gather(key = "sample", value = "count", -transcript_id)

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  filter(sample %in% selected_samples)

merged_transcriptome_counts <- merged_transcriptome_counts %>%
  rename(isoform = transcript_id)

merged_data <- merge(merged_transcriptome_clas, merged_transcriptome_counts, by = "isoform")

weighted_lengths <- merged_data %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * count) / sum(count))

final_data <- merge(weighted_lengths, noisy_reads_ratio_difference_pool2, by = "associated_gene")

correlation_result <- cor(final_data$weighted_length, final_data$noisy_reads_ratio_diff, method = "spearman")

print(correlation_result)
```




```{r, warning=FALSE}
locus_reference_df = read.table("/home/carblan/Documents/R_files/TFM/filtered_gtfs_with_loci/reference_annotation_loci.gtf", header = TRUE)

reference_clas = read.table("/home/carblan/Documents/R_files/TFM/reference_transcriptome_classification.txt", header = TRUE)

genes_files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes", pattern="B", full.names = TRUE)

colnames(locus_reference_df) <- c("chr", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")

extract_locus_transcript <- function(attribute) {
  locus <- sub(".*locus=([^;]+).*", "\\1", attribute)
  id <- sub(".*ID=([^;]+?)(;.*|$)", "\\1", attribute)
  return(data.frame(ID = id, locus = locus, stringsAsFactors=FALSE))
}

transcripts_data <- locus_reference_df %>%
  filter(feature == "transcript") %>%
  rowwise() %>%
	mutate(locus_transcript = list(extract_locus_transcript(attribute))) %>%
	unnest(locus_transcript) %>%
	select(ID, locus)

reference_clas <- reference_clas %>%
  left_join(transcripts_data, by = c("isoform" = "ID")) %>%
  filter(!(grepl("SIRV", chrom)))

reference_clas <- reference_clas %>% select(associated_gene, locus)

duplicated_genes <- reference_clas %>%
  group_by(associated_gene) %>%
  filter(n_distinct(locus) > 1) %>%
  ungroup()

if (nrow(duplicated_genes) > 0) {
  locus_gene_counts <- reference_clas %>%
    group_by(locus) %>%
    summarise(gene_count = n_distinct(associated_gene)) %>%
    ungroup()

  reference_clas <- reference_clas %>%
    left_join(locus_gene_counts, by = "locus")

  duplicated_genes <- duplicated_genes %>%
    left_join(locus_gene_counts, by = "locus")

  resolved_duplicates <- duplicated_genes %>%
    group_by(associated_gene) %>%
    summarise(
      chosen_locus = locus[which.max(gene_count)],
      .groups = "drop"
    )

  reference_clas <- reference_clas %>%
    left_join(resolved_duplicates, by = "associated_gene", suffix = c("", "_resolved")) %>%
    mutate(locus = ifelse(!is.na(chosen_locus), chosen_locus, locus)) %>%
    select(associated_gene, locus) %>%
    distinct()
}

add_locus_column <- function(sample_df, ref_df) {
  sample_df <- sample_df %>%
    left_join(ref_df, by = "associated_gene")
  return(sample_df)
}

samples_clas_files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications/", pattern = "B", full.names = TRUE)

for (sample_clas_file in samples_clas_files) {
  sample_clas <- read.table(sample_clas_file, header = TRUE)
  
  sample_clas <- sample_clas %>%
    filter(!(filter_result == "Artifact") & !(structural_category %in% c("intergenic", "antisense", "fusion", "genic_intron")))
  sample_clas <- add_locus_column(sample_clas, reference_clas)
  
  sample_name <- sub("_.*", "", basename(sample_clas_file))
  
  dir <- "/home/carblan/Documents/R_files/TFM/filtered_classifications_with_loci"
  if (!dir.exists(dir)) {
    dir.create(dir)
  }
  
  output_name <- paste0(sample_name, "_filtered_classification_with_loci.txt")
  output_path <- paste0("/home/carblan/Documents/R_files/TFM/filtered_classifications_with_loci/", output_name)
  write.table(sample_clas, file = output_path, sep = "\t", row.names = FALSE)
  
  final_df <- sample_clas %>%
    group_by(locus, chrom) %>%
    summarise(median_length = median(length),
              median_exons = median(exons),
              full_splice_match_counts = sum(FL[structural_category == "full-splice_match"]),
              incomplete_splice_match_counts = sum(FL[structural_category == "incomplete-splice_match"]),
              novel_in_catalog_counts = sum(FL[structural_category == "novel_in_catalog"]),
              novel_not_in_catalog_counts = sum(FL[structural_category == "novel_not_in_catalog"]),
              genic_counts = sum(FL[structural_category == "genic"]),
              total_counts = sum(FL))
  
  dir <- "/home/carblan/Documents/R_files/TFM/filtered_per_locus_classifications"
  if (!dir.exists(dir)) {
    dir.create(dir)
  }  
  
  output_name <- paste0(sample_name, "_per_locus_filtered_classification.txt")
  write.table(final_df, file = paste0("/home/carblan/Documents/R_files/TFM/filtered_per_locus_classifications/", output_name), sep = "\t", row.names = FALSE)
  
  genes_file <- genes_files[grepl(sample_name, genes_files)]
  genes_df <- read.table(genes_file, header = TRUE)
  genes <- genes_df$associated_gene
  
  sample_clas <- sample_clas %>%
    filter(associated_gene %in% genes)
  
  final_df <- sample_clas %>%
    group_by(locus, chrom) %>%
    summarise(median_length = median(length),
              median_exons = median(exons),
              full_splice_match_counts = sum(FL[structural_category == "full-splice_match"]),
              incomplete_splice_match_counts = sum(FL[structural_category == "incomplete-splice_match"]),
              novel_in_catalog_counts = sum(FL[structural_category == "novel_in_catalog"]),
              novel_not_in_catalog_counts = sum(FL[structural_category == "novel_not_in_catalog"]),
              genic_counts = sum(FL[structural_category == "genic"]),
              total_counts = sum(FL))
  
  dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci"
  if (!dir.exists(dir)) {
    dir.create(dir)
  }
  
  output_name <- paste0(sample_name, "_per_locus_filtered_classification.txt")
  write.table(final_df, file = paste0("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/", output_name), sep = "\t", row.names = FALSE)  
}

p <- read.table("/home/carblan/Documents/R_files/TFM/filtered_classifications_with_loci/B154_filtered_classification_with_loci.txt", header = TRUE)
```


```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-5]")]
old_files <- files[str_detect(basename(files), "^B15[1-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(locus, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  
          	.groups = "drop") %>%
	arrange(locus)  
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/age_groups_per_locus_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/age_groups_per_locus_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	locus = data$locus,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "locus", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	locus = combined_data$locus,
	chrom = combined_data$chrom_old,
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_locus.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Locus count", title = paste("Distribution per-locus noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM\n", "Loci in the intersection of both pools (but only pool 2 read counts)")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.1, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_noisy_reads_ratio/noisy_reads_ratio_difference_per_locus.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
```



```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool1 <- files[str_detect(basename(files), "^B3[1-3]")]
old_files_pool1 <- files[str_detect(basename(files), "^B15[1-2]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(locus, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(locus)  # Ordenar primero por chrom y luego por locus
 
  return(data_mean)
}

young_mean_ratios_pool1 <- calculate_mean_ratios(young_files_pool1)

old_mean_ratios_pool1 <- calculate_mean_ratios(old_files_pool1)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	locus = data$locus,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(young_mean_ratios_pool1)
old_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(old_mean_ratios_pool1)


calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "locus", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	locus = combined_data$locus,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_pool1 <- calculate_difference(young_noisy_reads_ratios_pool1, old_noisy_reads_ratios_pool1)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Locus count", title = paste("Distribution per-locus noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM\n", "Loci in the intersection of both pools (but only pool 2 read counts)")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.25, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.25, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(noisy_reads_ratio_difference_pool1)
print(histogram)
```


```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_loci/per_locus_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool2 <- files[str_detect(basename(files), "^B3[4-5]")]
old_files_pool2 <- files[str_detect(basename(files), "^B15[3-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(locus, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(locus)  # Ordenar primero por chrom y luego por locus
 
  return(data_mean)
}

young_mean_ratios_pool2 <- calculate_mean_ratios(young_files_pool2)

old_mean_ratios_pool2 <- calculate_mean_ratios(old_files_pool2)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	locus = data$locus,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(young_mean_ratios_pool2)
old_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(old_mean_ratios_pool2)


calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "locus", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	locus = combined_data$locus,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_pool2 <- calculate_difference(young_noisy_reads_ratios_pool2, old_noisy_reads_ratios_pool2)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Locus count", title = paste("Distribution per-locus noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM\n", "Loci in the intersection of both pools (but only pool 2 read counts)")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.25, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.25, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(noisy_reads_ratio_difference_pool2)
print(histogram)
```




```{r}
df_combined <- noisy_reads_ratio_difference_pool1 %>%
  inner_join(noisy_reads_ratio_difference_pool2, by = "associated_gene")

increased_noise_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x < 0 & noisy_reads_ratio_diff.y > 0)

increased_to_0_noise_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x < 0 & noisy_reads_ratio_diff.y == 0)

decreased_noise_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x > 0 & noisy_reads_ratio_diff.y < 0)

decreased_to_0_noise_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x > 0 & noisy_reads_ratio_diff.y == 0)

zero_to_positive_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x == 0 & noisy_reads_ratio_diff.y > 0)

zero_to_negative_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x == 0 & noisy_reads_ratio_diff.y < 0)

stable_0_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x == 0 & noisy_reads_ratio_diff.y == 0)

stable_positive_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x > 0 & noisy_reads_ratio_diff.y > 0)

stable_negative_genes <- df_combined %>%
  filter(noisy_reads_ratio_diff.x < 0 & noisy_reads_ratio_diff.y < 0)


cat("Genes que aumentaron de negativo a positivo:\n")
cat("\nTotal:", nrow(increased_noise_genes), "\n")

cat("\nGenes que aumentaron de negativo a 0:\n")
cat("\nTotal:", nrow(increased_to_0_noise_genes), "\n")

cat("\nGenes que disminuyeron de positivo a negativo:\n")
cat("\nTotal:", nrow(decreased_noise_genes), "\n")
cat("\nTop 15 genes que más cambiaron al bajar de positivo a negativo:\n")
decreased_noise_genes$difference <- decreased_noise_genes$noisy_reads_ratio_diff.x - decreased_noise_genes$noisy_reads_ratio_diff.y
decreased_noise_genes_sorted <- decreased_noise_genes[order(-decreased_noise_genes$difference), ]
top_15_genes <- decreased_noise_genes_sorted[1:15, "associated_gene"]
print(top_15_genes)


cat("\nGenes que disminuyeron de positivo a 0:\n")
cat("\nTotal:", nrow(decreased_to_0_noise_genes), "\n")

cat("\nGenes que tenían ruido 0 en el primer df y mayor a 0 en el segundo df:\n")
cat("\nTotal:", nrow(zero_to_positive_genes), "\n")

cat("\nGenes que tenían ruido 0 en el primer df y menor a 0 en el segundo df:\n")
cat("\nTotal:", nrow(zero_to_negative_genes), "\n")

cat("\nGenes que se mantuvieron en 0:\n")
cat("\nTotal:", nrow(stable_0_genes), "\n")

cat("\nGenes que se mantuvieron en valores positivos:\n")
cat("\nTotal:", nrow(stable_positive_genes), "\n")

cat("\nGenes que se mantuvieron en valores negativos:\n")
cat("\nTotal:", nrow(stable_negative_genes), "\n")
```




```{r}
df <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", head = TRUE)

df <- df[order(df$noisy_reads_ratio_diff, decreasing = TRUE), ]

gene_list <- df$noisy_reads_ratio_diff
names(gene_list) <- df$associated_gene

gsea_result <- gseGO(geneList = gene_list,
                     OrgDb = org.Mm.eg.db,
                     keyType = "SYMBOL",
                     ont = "BP",
                     nPerm = 10000,
                     minGSSize = 1,
                     maxGSSize = 250,
                     pvalueCutoff = 0.8,
                     pAdjustMethod = "BH",
                     verbose = TRUE)


dotplot(gsea_result, showCategory=10, split=".sign") + facet_grid(.~.sign) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10), 
        axis.text.y = element_text(size = 5),
        strip.text = element_text(size = 10),
        plot.title = element_text(size = 14, face = "bold")) +
  ggtitle("GSEA dotplot (adjusted p-value)")

```


```{r}
df_pool1 <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_1_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", head = TRUE)


df_pool1 <- df_pool1[order(df_pool1$noisy_reads_ratio_diff, decreasing = TRUE), ]


gene_list_pool1 <- df_pool1$noisy_reads_ratio_diff
names(gene_list_pool1) <- df_pool1$associated_gene


gsea_result <- gseGO(geneList = gene_list_pool1,
                     OrgDb = org.Mm.eg.db,
                     keyType = "SYMBOL",
                     ont = "BP",
                     nPerm = 10000,
                     minGSSize = 1,
                     maxGSSize = 500,
                     pvalueCutoff = 0.8,
                     pAdjustMethod = "BH",
                     verbose = TRUE)

dotplot(gsea_result, showCategory=10, split=".sign") + facet_grid(.~.sign) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10), 
        axis.text.y = element_text(size = 5),
        strip.text = element_text(size = 10),
        plot.title = element_text(size = 14, face = "bold")) +
  ggtitle("GSEA dotplot (adjusted p-value)")
```


```{r}
df_pool2 <- read.table("/home/carblan/Documents/R_files/TFM/FSM>=10_pool_2_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", head = TRUE)

df_pool2 <- df_pool2[order(df_pool2$noisy_reads_ratio_diff, decreasing = TRUE), ]

gene_list_pool2 <- df_pool2$noisy_reads_ratio_diff
names(gene_list_pool2) <- df_pool2$associated_gene

gsea_result <- gseGO(geneList = gene_list_pool2,
                     OrgDb = org.Mm.eg.db,
                     keyType = "SYMBOL",
                     ont = "ALL",
                     nPerm = 10000,
                     minGSSize = 3,
                     maxGSSize = 800,
                     pvalueCutoff = 0.8,
                     pAdjustMethod = "BH",
                     verbose = TRUE)

dotplot(gsea_result, showCategory=10, split=".sign") + facet_grid(.~.sign) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10), 
        axis.text.y = element_text(size = 5),
        strip.text = element_text(size = 10),
        plot.title = element_text(size = 14, face = "bold")) +
  ggtitle("GSEA dotplot (adjusted p-value)")
```




```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/alejandro_filtered_classifications", pattern = "^B3[1-5]|^B15[1-4]", full.names = TRUE)

process_file <- function(file) {
  data <- read.table(file, header = TRUE)
 
  data_filtered <- data %>%
	filter(!(grepl("SIRV", chrom)) & !(filter_result == "Artifact") & !(structural_category %in% c("intergenic", "antisense", "fusion", "genic_intron")))
 
  inter_df <- data_filtered %>%
	select(isoform, chrom, associated_gene, structural_category, FL, length, exons) %>%
	arrange(associated_gene)  # Ordenar por associated_gene
 
  final_df <- inter_df %>%
	group_by(associated_gene, chrom) %>%
	summarise(median_length = median(length),
          	median_exons = median(exons),
          	full_splice_match_counts = sum(FL[structural_category == "full-splice_match"]),
          	incomplete_splice_match_counts = sum(FL[structural_category == "incomplete-splice_match"]),
          	novel_in_catalog_counts = sum(FL[structural_category == "novel_in_catalog"]),
          	novel_not_in_catalog_counts = sum(FL[structural_category == "novel_not_in_catalog"]),
          	genic_counts = sum(FL[structural_category == "genic"]),
          	total_counts = sum(FL))
 
  sample_name <- sub("_.*", "", basename(file))
  output_name <- paste0(sample_name, "_per_gene_filtered_classification.txt")
  write.table(final_df, file = paste0("/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications/", output_name), sep = "\t", row.names = FALSE)
}

dir <- "/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in files) {
  process_file(file)
}
```


```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications",
                	full.names = TRUE)
filtered_data <- list()

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-5]")]
old_files <- files[str_detect(basename(files), "^B15[1-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```

```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.005, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM\n", "pools intersection, isoseq pipeline")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 1000, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.1, y = 1000, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
noisy_reads_ratio_difference <- read.table("/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(noisy_reads_ratio_difference)
print(histogram)
```


```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications",
                	full.names = TRUE)
filtered_data <- list()

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool1 <- files[str_detect(basename(files), "^B3[1-3]")]
old_files_pool1 <- files[str_detect(basename(files), "^B15[1-2]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios_pool1 <- calculate_mean_ratios(young_files_pool1)

old_mean_ratios_pool1 <- calculate_mean_ratios(old_files_pool1)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(young_mean_ratios_pool1)
old_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(old_mean_ratios_pool1)


calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_pool1 <- calculate_difference(young_noisy_reads_ratios_pool1, old_noisy_reads_ratios_pool1)
```


```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.005, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM\n", "pools intersection, pool 1 counts, isoseq pipeline")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.2, y = 700, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.2, y = 700, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
histogram <- RBN_histogram(noisy_reads_ratio_difference_pool1)
print(histogram)
```



```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications",
                	full.names = TRUE)
filtered_data <- list()

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_intersection_filtered_genes/per_gene_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool2 <- files[str_detect(basename(files), "^B3[4-5]")]
old_files_pool2 <- files[str_detect(basename(files), "^B15[3-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios_pool2 <- calculate_mean_ratios(young_files_pool2)

old_mean_ratios_pool2 <- calculate_mean_ratios(old_files_pool2)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(young_mean_ratios_pool2)
old_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(old_mean_ratios_pool2)


calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_pool2 <- calculate_difference(young_noisy_reads_ratios_pool2, old_noisy_reads_ratios_pool2)
```


```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.005, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM\n", "pool 2, isoseq pipeline")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.2, y = 700, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.2, y = 700, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
histogram <- RBN_histogram(noisy_reads_ratio_difference_pool2)
print(histogram)
```



```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications",
                	full.names = TRUE)
pool_1_files <- files[str_detect(basename(files), "^B3[1-3]|^B15[1-2]")]
filtered_data <- list()

for (file in pool_1_files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in pool_1_files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-3]")]
old_files <- files[str_detect(basename(files), "^B15[1-2]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```

```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_1_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
```



```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_filtered_per_gene_classifications",
                	full.names = TRUE)
pool_2_files <- files[str_detect(basename(files), "^B3[4-5]|^B15[3-4]")]
filtered_data <- list()

for (file in pool_2_files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  data <- data[data$full_splice_match_counts >= 10,]
 
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in pool_2_files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-5]")]
old_files <- files[str_detect(basename(files), "^B15[1-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,  # Puedes elegir la columna chrom de old_data o young_data
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```

```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.01, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young:\n", "NNC + NIC / NNC + NIC + FSM")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.25, y = 1250, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.25, y = 1250, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/isoseq_FSM>=10_pool_2_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
```




Once we have the per gene classifications with the counts, median length and median number of exons, we will calculate the noisy reads ratio difference per gene, alongside its length and number of exons. For this we first filter the genes that have at least 10 reads assigned as FSM in every sample, and also appear in every sample (to avoid genes that are underrepresented) Then, we obtain the ratio for each structural category of each gene, then the mean of the ratio of each structural category for the samples in the young group of mice and the samples in the old group of mice, independently. Then, the noisy reads ratio (NIC+NNC/NIC+NNC+FSM) is calculated for each gene of each age group and the difference (old - young) is obtained.

For the length and number of exons of the gene, we get the median length and median number of exons of each gene for each sample and compute the average for each age group. Then, from that average for each age group, a second average is calculated to obtain the final length and number of exons per gene.

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_per_gene_classifications",
                	full.names = TRUE)
filtered_data <- list()

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
  filtered_data[[file]] <- data
}

associated_genes <- Reduce(intersect, lapply(filtered_data, function(x) x$associated_gene))

dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes"
if (!dir.exists(dir)) {
  dir.create(dir)
}

for (file in files) {
  data <- filtered_data[[file]]
  data <- data[data$associated_gene %in% associated_genes,]
 
  write.table(data, file.path(dir, basename(file)), sep="\t", row.names=FALSE)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes", pattern=".txt", full.names = TRUE)

output_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/per_gene_structural_categories_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (file in files) {
  data <- read.table(file, header=TRUE, stringsAsFactors=FALSE)
 
  categories <- colnames(data)[5:ncol(data)]
  for (category in categories) {
	data[[category]] <- data[[category]] / data$total_counts
  }
 
  output_file <- file.path(output_dir, basename(file))
  write.table(data, output_file, sep="\t", quote=FALSE, row.names=FALSE)
}

input_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/per_gene_structural_categories_ratios"

files <- list.files(path = input_dir, pattern=".txt", full.names = TRUE)

young_files <- files[str_detect(basename(files), "^B3[1-5]")]
old_files <- files[str_detect(basename(files), "^B15[1-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  
          	.groups = "drop") %>%
	arrange(associated_gene)  
 
  return(data_mean)
}

young_mean_ratios <- calculate_mean_ratios(young_files)

old_mean_ratios <- calculate_mean_ratios(old_files)

output_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/age_groups_per_gene_categories_mean_ratios"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

write.table(young_mean_ratios, file.path(output_dir, "young_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_mean_ratios, file.path(output_dir, "old_mean_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/age_groups_per_gene_categories_mean_ratios"

young_mean_ratios <- read.table(file.path(input_dir, "young_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_mean_ratios <- read.table(file.path(input_dir, "old_mean_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios <- calculate_noisy_reads_ratios(young_mean_ratios)
old_noisy_reads_ratios <- calculate_noisy_reads_ratios(old_mean_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(young_noisy_reads_ratios, file.path(output_dir, "young_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)
write.table(old_noisy_reads_ratios, file.path(output_dir, "old_noisy_reads_ratios.txt"), sep="\t", quote=FALSE, row.names=FALSE)

input_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/per_gene_noisy_reads_ratio"

young_noisy_reads_ratios <- read.table(file.path(input_dir, "young_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)
old_noisy_reads_ratios <- read.table(file.path(input_dir, "old_noisy_reads_ratios.txt"), header=TRUE, stringsAsFactors=FALSE)

calculate_difference <- function(young_data, old_data) {
  combined_data <- merge(old_data, young_data, by = "associated_gene", suffixes = c("_old", "_young"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old - combined_data$noisy_reads_ratio_young

  media_average_length <- rowMeans(combined_data[, c("average_length_old", "average_length_young")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old", "average_exons_young")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old,
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference <- calculate_difference(young_noisy_reads_ratios, old_noisy_reads_ratios)

output_dir <- "/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/per_gene_noisy_reads_ratio"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}

write.table(noisy_reads_ratio_difference, file.path(output_dir, "noisy_reads_ratio_difference_per_gene.txt"), sep="\t", quote=FALSE, row.names=FALSE)

```


```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.025, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old vs. young")) +
    scale_fill_manual(name = "Greater in old", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.1, y = 1000, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 5) +
    annotate("text", x = 0.1, y = 1000, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 5) +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(histograma)
}
```

```{r}
data <- read.table("/home/carblan/Documents/R_files/TFM/all_samples_intersection_filtered_genes/per_gene_noisy_reads_ratio/noisy_reads_ratio_difference_per_gene.txt", header = TRUE)
histogram <- RBN_histogram(data)
print(histogram)
```



```{r, message = FALSE}
files <- list.files("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios", full.names = TRUE)

process_files <- function(file) {
  df <- read.table(file, header = TRUE)
  df$rbn <- (df$novel_in_catalog_counts + df$novel_not_in_catalog_counts) / (df$novel_in_catalog_counts + df$novel_not_in_catalog_counts + df$full_splice_match_counts)
  noise_df <- df[, c("associated_gene",	"chrom", "rbn")]
  sample_name <- sub("_.*", "", basename(file))
  noise_df$sample <- sub("_.*", "", basename(file))
  output_name <- paste0(sample_name, "_per_gene_rbn.txt")
  write.table(noise_df, file = paste0("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio/", output_name), sep = "\t", row.names = FALSE)
  return(noise_df)
}

rbn_files <- lapply(files, process_files)

combined_df <- bind_rows(rbn_files)
combined_df$group <- ifelse(combined_df$sample %in% c("B151", "B152", "B153", "B154"), "old", "young")
combined_df$pool <- ifelse(combined_df$sample %in% c("B31", "B32", "B33", "B151", "B152"), "pool 1", "pool 2")
samples_order <- c("B31", "B32", "B33", "B34", "B35", "B151", "B152", "B153", "B154")
combined_df$sample <- factor(combined_df$sample, levels = samples_order)
combined_df$group <- factor(combined_df$group, levels = c("young", "old"))

p <- ggplot(combined_df, aes(x = sample, y = rbn, fill = group, color = group)) +
  geom_violin(alpha = 0.3) +
  geom_boxplot(alpha = 0.5, width = 0.5, outlier.shape = NA) + 
  scale_fill_conesa(palette = "main") +
  scale_color_conesa(palette = "main") +
  coord_cartesian(ylim = c(0, 0.25)) +
  labs(x = "Sample", y = "RNA biological noise") +
  theme_minimal() +
  my_theme

ggsave("pacbio_RBN_capture.png", p, path = "/home/carblan/Documents/R_files/TFM/figures/RBNcapture/ONT", width = 10, height = 7)

library(ggpubr)
p <- ggplot(combined_df, aes(x = group, y = rbn, fill = group, color = group)) +
  geom_boxplot(alpha = 0.5) +
  theme_minimal() +
  my_theme +
  labs(x = "Age group", y = "RBN", fill = "Age group", color = "Age group") +
  scale_fill_conesa(palette = "main") +
  scale_color_conesa(palette = "main") +
  coord_cartesian(ylim = c(0, 0.25)) +
  stat_compare_means( aes(label = ..p.signif..), 
                        label.x = 1.5, label.y = 0.2, hide.ns = TRUE, method = "t.test", show.legend=FALSE)

ggsave("nanopore_RBN_capture_test.png", p, path = "/home/carblan/Documents/R_files/TFM/figures/RBNcapture/ONT")
```

```{r}
files <- list.files("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio", pattern = "B", full.names = TRUE)

total_rbn <- function(file) {
  df <- read.table(file, header = TRUE)
  total_rbn <- sum(df$rbn)
  return(total_rbn)
}

results <- sapply(files, total_rbn, USE.NAMES = TRUE)
samples <- sapply(strsplit(basename(files), "_"), `[`, 1)
rbn_df <- data.frame(total_rbn = results, row.names = samples)

samples_order <- c("B31", "B32", "B33", "B151", "B152", "B34", "B35", "B153", "B154")
rbn_df <- rbn_df[samples_order, , drop = FALSE]

rbn_df$sample <- rownames(rbn_df)
rbn_df$group <- ifelse(rbn_df$sample %in% c("B151", "B152", "B153", "B154"), "old", "young")
rbn_df$pool <- ifelse(rbn_df$sample %in% c("B31", "B32", "B33", "B151", "B152"), "pool 1", "pool 2")

rbn_df$sample <- factor(rbn_df$sample, levels = samples_order)

library(ggplot2)

ggplot(rbn_df, aes(x = sample, y = total_rbn, fill = group, color = group)) +
  geom_bar(stat = "identity", alpha = 0.7) +
  scale_fill_conesa(palette = "main") +
  scale_color_conesa(palette = "main") +
  theme_minimal() +
  labs(title = "Per sample total RBN", x = "Sample", y = "Total RBN") +
  theme(
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 14),
    legend.text = element_text(size = 12)
  ) +
  ylim(0, 800)
```



We can also compare noise between pools within the same age group to actually ascertain how pool enrichment and the age group interact. 

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios", pattern=".txt", full.names = TRUE)

young_files_pool1 <- files[str_detect(basename(files), "^B3[1-3]")]
young_files_pool2 <- files[str_detect(basename(files), "^B3[4-5]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

young_mean_ratios_pool1 <- calculate_mean_ratios(young_files_pool1)

young_mean_ratios_pool2 <- calculate_mean_ratios(young_files_pool2)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

young_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(young_mean_ratios_pool1)
young_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(young_mean_ratios_pool2)


calculate_difference <- function(young_data_pool1, young_data_pool2) {
  combined_data <- merge(young_data_pool1, young_data_pool2, by = "associated_gene", suffixes = c("_young_pool1", "_young_pool2"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_young_pool1 - combined_data$noisy_reads_ratio_young_pool2

  media_average_length <- rowMeans(combined_data[, c("average_length_young_pool1", "average_length_young_pool2")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_young_pool1", "average_exons_young_pool2")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_young_pool2,  # Puedes elegir la columna chrom de young_data_pool1 o young_data_pool2
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_young <- calculate_difference(young_noisy_reads_ratios_pool1, young_noisy_reads_ratios_pool2)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.025, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference young (pool 1) vs. young (pool 2):\n", "Genes in the intersection of both pools and age groups")) +
    scale_fill_manual(name = "Greater in pool 1", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.25, y = 25, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 12) +
    annotate("text", x = 0.25, y = 25, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 12) +
    theme(
      axis.text.x = element_text(size = 17),
      axis.text.y = element_text(size = 17),
      axis.title.x = element_text(size = 20),
      axis.title.y = element_text(size = 20),
      plot.title = element_text(size = 25, hjust = 0.5),
      legend.title = element_text(size = 17),
      legend.text = element_text(size = 15))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(noisy_reads_ratio_difference_young)
print(histogram)
```


```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B3", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)

combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B31", "B32", "B33"), "pool 1", "pool 2")

weighted_length_df_young <- combined_df %>%
  group_by(sample, associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL), .groups = 'drop')

weighted_length_df <- merge(noisy_reads_ratio_difference_young, weighted_length_df_young, by = "associated_gene")

top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_pearson, "\n")

ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "young samples, all transcript models"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))

```



```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}


files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B3", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B31", "B32", "B33"), "pool 1", "pool 2")

combined_df_young1 <- combined_df %>% filter(sample %in% c("B31", "B32", "B33"))
combined_df_young2 <- combined_df %>% filter(sample %in% c("B34", "B35"))

weighted_length_df_young1 <- combined_df_young1 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(pool = "pool 1")

weighted_length_df_young2 <- combined_df_young2 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(pool = "pool 2")

weighted_length_df <- merge(noisy_reads_ratio_difference_young, weighted_length_df_young1, by = "associated_gene")

top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_pearson, "\n")

ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "young samples, pool 1 transcripts for weigthed length calculation"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))



top_10_percent_genes <- weighted_length_df %>%
  filter(top_10_percent == "Top 10%")

correlation_spearman <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_pearson, "\n")

ggplot(top_10_percent_genes, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "young samples, pool 1 transcripts for weigthed length calculation, top 10% genes with hight diff"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(top_10_percent_genes$noisy_reads_ratio_diff), y = max(top_10_percent_genes$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))


```



```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B3", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B31", "B32", "B33"), "pool 1", "pool 2")


combined_df_young1 <- combined_df %>% filter(sample %in% c("B31", "B32", "B33"))
combined_df_young2 <- combined_df %>% filter(sample %in% c("B34", "B35"))

weighted_length_df_young1 <- combined_df_young1 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length_pool1 = sum(length * FL) / sum(FL))

weighted_length_df_young2 <- combined_df_young2 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length_pool2 = sum(length * FL) / sum(FL))

weighted_length_df <- merge(weighted_length_df_young1, weighted_length_df_young2, by = "associated_gene")

weighted_length_df <- weighted_length_df %>%
  mutate(weighted_length_diff = weighted_length_pool1 - weighted_length_pool2)

weighted_length_df <- merge(noisy_reads_ratio_difference_young, weighted_length_df, by = "associated_gene")

top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length_diff, method = "spearman")
cat("Weighted length of genes (transcript models from young samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length_diff, method = "pearson")
cat("Weighted length of genes (transcript models from young samples) and noise difference correlation (Pearson):\n", correlation_pearson, "\n")

ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length_diff, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = "Correlation between Weighted Length Difference and Noisy Reads Ratio Difference (young)",
       x = "Noisy Reads Ratio Difference",
       y = "Weighted Length Difference") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length_diff), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))
```



```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B3", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)

combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B31", "B32", "B33"), "pool 1", "pool 2")

rbn_files <- list.files("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio", pattern = "B3", full.names = TRUE)

rbn_df <- list()
for (rbn_file in rbn_files) {
  df <- read.table(rbn_file, header = TRUE)
  sample_name <- sub("_.*", "", basename(rbn_file))
  rbn_df[[sample_name]] <- df 
}

rbn_df <- bind_rows(rbn_df)
rbn_df <- rbn_df %>%
  group_by(associated_gene) %>%
  summarise(average_rbn = mean(rbn, na.rm = TRUE))

rbn_df <- merge(rbn_df, noisy_reads_ratio_difference_young, by = "associated_gene")
top_10_percent_threshold <- quantile(rbn_df$noisy_reads_ratio_diff, 0.9)
rbn_df <- rbn_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))


ggplot(rbn_df, aes(x = average_rbn, y = noisy_reads_ratio_diff, color = top_10_percent)) +
  geom_point(alpha = 0.5) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  labs(title = "MA Plot (young)",
       x = "Average RBN",
       y = "Noisy Reads Ratio Difference") +
  scale_color_conesa(palette = "main") +
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))

```



We can also compare noise between pools within the same age group to actually ascertain how pool enrichment and the age group interact. 

```{r}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/FSM>1_intersection_filtered_genes_nanopore_not_subsampled/per_gene_structural_categories_ratios", pattern=".txt", full.names = TRUE)

old_files_pool1 <- files[str_detect(basename(files), "^B15[1-2]")]
old_files_pool2 <- files[str_detect(basename(files), "^B15[3-4]")]

calculate_mean_ratios <- function(files) {
  data <- map_dfr(files, ~read.table(.x, header=TRUE, stringsAsFactors=FALSE))
 
  data_mean <- data %>%
	group_by(associated_gene, chrom) %>%
	summarise(across(starts_with("median_length"), mean),
          	across(starts_with("median_exons"), mean),
          	across(starts_with("full"), mean),
          	across(starts_with("incomplete"), mean),
          	across(starts_with("novel_not"), mean),
          	across(starts_with("novel_in"), mean),
          	across(starts_with("genic"), mean),
          	total_counts = mean(total_counts),  # Calcular la media de la columna "total_counts"
          	.groups = "drop") %>%
	arrange(associated_gene)  # Ordenar primero por chrom y luego por associated_gene
 
  return(data_mean)
}

old_mean_ratios_pool1 <- calculate_mean_ratios(old_files_pool1)

old_mean_ratios_pool2 <- calculate_mean_ratios(old_files_pool2)

calculate_noisy_reads_ratios <- function(data) {
  noisy_reads_ratio <- (data$novel_in_catalog_counts + data$novel_not_in_catalog_counts) / (data$full_splice_match_counts + data$novel_in_catalog_counts + data$novel_not_in_catalog_counts)
 
  noisy_reads_ratios <- data.frame(
	associated_gene = data$associated_gene,
	chrom = data$chrom,
	average_length = data$median_length,
	average_exons = data$median_exons,
	noisy_reads_ratio = noisy_reads_ratio
  )
 
  return(noisy_reads_ratios)
}

old_noisy_reads_ratios_pool1 <- calculate_noisy_reads_ratios(old_mean_ratios_pool1)
old_noisy_reads_ratios_pool2 <- calculate_noisy_reads_ratios(old_mean_ratios_pool2)


calculate_difference <- function(old_data_pool1, old_data_pool2) {
  combined_data <- merge(old_data_pool1, old_data_pool2, by = "associated_gene", suffixes = c("_old_pool1", "_old_pool2"))

  noisy_reads_ratio_diff <- combined_data$noisy_reads_ratio_old_pool1 - combined_data$noisy_reads_ratio_old_pool2

  media_average_length <- rowMeans(combined_data[, c("average_length_old_pool1", "average_length_old_pool2")], na.rm = TRUE)
  media_average_exons <- rowMeans(combined_data[, c("average_exons_old_pool1", "average_exons_old_pool2")], na.rm = TRUE)

  noisy_reads_ratio_difference <- data.frame(
	associated_gene = combined_data$associated_gene,
	chrom = combined_data$chrom_old_pool2,  # Puedes elegir la columna chrom de old_data_pool1 o old_data_pool2
	average_length = media_average_length,
	average_exons = media_average_exons,
	noisy_reads_ratio_diff = noisy_reads_ratio_diff
  )

  return(noisy_reads_ratio_difference)
}

noisy_reads_ratio_difference_old <- calculate_difference(old_noisy_reads_ratios_pool1, old_noisy_reads_ratios_pool2)
```



```{r}
RBN_histogram <- function(data) {
  data_filtrado <- data[complete.cases(data) & data$noisy_reads_ratio_diff != 0, ]
  
  genes_above_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] > 0)
  genes_below_0 <- sum(data_filtrado[, "noisy_reads_ratio_diff"] < 0)
  total_genes <- nrow(data_filtrado)
  porcentaje_above_0 <- round((genes_above_0 / total_genes) * 100, 2)
  porcentaje_below_0 <- round((genes_below_0 / total_genes) * 100, 2)
  
  histograma <- ggplot(data_filtrado, aes(x = data_filtrado[, "noisy_reads_ratio_diff"], fill = ifelse(data_filtrado[, "noisy_reads_ratio_diff"] > 0, "above_0", "below_0"))) +
    geom_histogram(binwidth = 0.025, boundary = 0, color = "black") +
    geom_vline(xintercept = 0, color = "darkblue", linetype = "dashed") +
    labs(x = "Ratio difference", y = "Gene count", title = paste("Distribution per-gene noisy reads ratio difference old (pool 1) vs. old (pool 2):\n", "Genes in the intersection of both pools and age groups")) +
    scale_fill_manual(name = "Greater in pool 1", values = c("above_0" = "#aae56c", "below_0" = "#ff5151"), labels = c("Yes", "No")) +
    coord_cartesian(xlim = c(-max(abs(data_filtrado[, "noisy_reads_ratio_diff"])), max(abs(data_filtrado[, "noisy_reads_ratio_diff"])))) +
    annotate("text", x = -0.25, y = 25, label = paste(porcentaje_below_0, "%"), color = "#ff5151", size = 12) +
    annotate("text", x = 0.25, y = 25, label = paste(porcentaje_above_0, "%"), color = "darkgreen", size = 12) +
    theme(
      axis.text.x = element_text(size = 17),
      axis.text.y = element_text(size = 17),
      axis.title.x = element_text(size = 20),
      axis.title.y = element_text(size = 20),
      plot.title = element_text(size = 25, hjust = 0.5),
      legend.title = element_text(size = 17),
      legend.text = element_text(size = 15))
  
  return(histograma)
}
```

```{r}
histogram <- RBN_histogram(noisy_reads_ratio_difference_old)
print(histogram)
```


```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B15", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B151", "B152"), "pool 1", "pool 2")

weighted_length_df_old <- combined_df %>%
  group_by(sample, associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL), .groups = 'drop')

weighted_length_df <- merge(noisy_reads_ratio_difference_old, weighted_length_df_old, by = "associated_gene")

top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation, "\n")

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation, "\n")


ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = "Correlation between Weighted Length Difference and Noisy Reads Ratio Difference (old)",
       x = "Noisy Reads Ratio Difference",
       y = "Weighted Length Difference") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))
```



```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B15", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B151", "B152"), "pool 1", "pool 2")

combined_df_old1 <- combined_df %>% filter(sample %in% c("B151", "B152"))
combined_df_old2 <- combined_df %>% filter(sample %in% c("B153", "B154"))

weighted_length_df_old1 <- combined_df_old1 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(pool = "pool 1")

weighted_length_df_old2 <- combined_df_old2 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length = sum(length * FL) / sum(FL)) %>%
  mutate(pool = "pool 2")

weighted_length_df <- merge(noisy_reads_ratio_difference_old, weighted_length_df_old1, by = "associated_gene")

top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_pearson, "\n")

ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "old samples, pool 1 transcripts for weigthed length calculation"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))



top_10_percent_genes <- weighted_length_df %>%
  filter(top_10_percent == "Top 10%")

correlation_spearman <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "spearman")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(top_10_percent_genes$noisy_reads_ratio_diff, top_10_percent_genes$weighted_length, method = "pearson")
cat("Weighted length of genes (transcript models froms old samples) and noise difference correlation (Spearman):\n", correlation_pearson, "\n")

ggplot(top_10_percent_genes, aes(x = noisy_reads_ratio_diff, y = weighted_length, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = paste("Correlation between Gene Weighted Length and Noisy Reads Ratio Difference\n", "old samples, pool 1 transcripts for weigthed length calculation, top 10% genes with hight diff"),
       x = "Noisy Reads Ratio Difference",
       y = " Gene Weighted Length") +
  annotate("text", x = min(top_10_percent_genes$noisy_reads_ratio_diff), y = max(top_10_percent_genes$weighted_length), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))


```


```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B15", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)


combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B151", "B152"), "pool 1", "pool 2")


combined_df_old1 <- combined_df %>% filter(sample %in% c("B151", "B152"))
combined_df_old2 <- combined_df %>% filter(sample %in% c("B153", "B154"))

weighted_length_df_old1 <- combined_df_old1 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length_pool1 = sum(length * FL) / sum(FL))

weighted_length_df_old2 <- combined_df_old2 %>%
  group_by(associated_gene) %>%
  summarise(weighted_length_pool2 = sum(length * FL) / sum(FL))

weighted_length_df <- merge(weighted_length_df_old1, weighted_length_df_old2, by = "associated_gene")

weighted_length_df <- weighted_length_df %>%
  mutate(weighted_length_diff = weighted_length_pool1 - weighted_length_pool2)

weighted_length_df <- merge(noisy_reads_ratio_difference_old, weighted_length_df, by = "associated_gene")

top_10_percent_threshold <- quantile(weighted_length_df$noisy_reads_ratio_diff, 0.9)
weighted_length_df <- weighted_length_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

correlation_spearman <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length_diff, method = "spearman")
cat("Weighted length of genes (transcript models from old samples) and noise difference correlation (Spearman):\n", correlation_spearman, "\n")

correlation_pearson <- cor(weighted_length_df$noisy_reads_ratio_diff, weighted_length_df$weighted_length_diff, method = "pearson")
cat("Weighted length of genes (transcript models from old samples) and noise difference correlation (Pearson):\n", correlation_pearson, "\n")


ggplot(weighted_length_df, aes(x = noisy_reads_ratio_diff, y = weighted_length_diff, color = top_10_percent)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, color = "blue", linetype = "dashed") +
  labs(title = "Correlation between Weighted Length Difference and Noisy Reads Ratio Difference (old)",
       x = "Noisy Reads Ratio Difference",
       y = "Weighted Length Difference") +
  annotate("text", x = min(weighted_length_df$noisy_reads_ratio_diff), y = max(weighted_length_df$weighted_length_diff), 
           label = paste("Spearman: ", round(correlation_spearman, 3), "\nPearson: ", round(correlation_pearson, 3)), 
           hjust = 0, vjust = 1, size = 5, color = "red") +
  scale_color_conesa(palette = "main") + 
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))
```


```{r}
filter_dataframe <- function(file, associated_genes) {
  lines <- readLines(file)
  valid_lines <- lines[sapply(strsplit(lines, "\t"), length) == 49] # Al parecer hay alguna isoforma con la info incompleta de algún campo en el classification.
  df <- read.table(text = paste(valid_lines, collapse = "\n"), header = TRUE, sep = "\t", fill = TRUE, quote = "")
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}
files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications_nanopore", pattern = "B15", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)

combined_df <- bind_rows(filtered_dfs)
combined_df$pool <- ifelse(combined_df$sample %in% c("B151", "B152"), "pool 1", "pool 2")

rbn_files <- list.files("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes/per_gene_noisy_reads_ratio", pattern = "B151|B152", full.names = TRUE)

rbn_df <- list()
for (rbn_file in rbn_files) {
  df <- read.table(rbn_file, header = TRUE)
  sample_name <- sub("_.*", "", basename(rbn_file))
  rbn_df[[sample_name]] <- df 
}

rbn_df <- bind_rows(rbn_df)
rbn_df <- rbn_df %>%
  group_by(associated_gene) %>%
  summarise(average_rbn = mean(rbn, na.rm = TRUE))

rbn_df <- merge(rbn_df, noisy_reads_ratio_difference_old, by = "associated_gene")
top_10_percent_threshold <- quantile(rbn_df$noisy_reads_ratio_diff, 0.9)
rbn_df <- rbn_df %>%
  mutate(top_10_percent = ifelse(noisy_reads_ratio_diff >= top_10_percent_threshold, "Top 10%", "Rest"))

ggplot(rbn_df, aes(x = average_rbn, y = noisy_reads_ratio_diff, color = top_10_percent)) +
  geom_point(alpha = 0.5) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  labs(title = "MA Plot (old, pool 1 average RBN)",
       x = "Average RBN",
       y = "Noisy Reads Ratio Difference") +
  scale_color_conesa(palette = "main") +
  theme(
    axis.text.x = element_text(size = 17),
    axis.text.y = element_text(size = 17),
    axis.title.x = element_text(size = 20),
    axis.title.y = element_text(size = 20),
    plot.title = element_text(size = 25, hjust = 0.5),
    legend.title = element_text(size = 17),
    legend.text = element_text(size = 15))
"coord_fixed(ratio = 1) +  
xlim(0, 0.3) +
ylim(0, 0.3) +"


```



```{r}
filter_dataframe <- function(file, associated_genes) {
  df <- read.table(file, header = TRUE)
  df <- df[df$filter_result != "Artifact", ]
  df <- df[df$associated_gene %in% associated_genes, ]
  df$sample <- sub("_.*", "", basename(file))
  return(df)
}

files <- list.files(path = "/home/carblan/Documents/R_files/TFM/filtered_classifications", pattern = "B", full.names = TRUE)
filtered_dfs <- lapply(files, filter_dataframe, associated_genes)

combined_df <- bind_rows(filtered_dfs)
combined_df$age_group <- ifelse(combined_df$sample %in% c("B151", "B152", "B153", "B154"), "old", "young")
combined_df$pool <- ifelse(combined_df$sample %in% c("B31", "B32", "B33", "B151", "B152"), "pool 1", "pool 2")

rbn_files <- list.files("/home/carblan/Documents/R_files/TFM/FSM>=10_intersection_filtered_genes_nanopore_not_subsampled/per_gene_noisy_reads_ratio", pattern = "B", full.names = TRUE)

for (unique_sample in unique(combined_df$sample)) {
  combined_df_sample <- combined_df %>% filter(sample == unique_sample)
  weighted_length_df_sample <- combined_df_sample %>%
    group_by(associated_gene) %>%
    summarise(weighted_length = sum(length * FL) / sum(FL))
  rbn_file <- rbn_files[grepl(unique_sample, rbn_files)]
  rbn_df_sample <- read.table(rbn_file, header = TRUE)
  weighted_length_df <- merge(rbn_df_sample, weighted_length_df_sample, by = "associated_gene")
  correlation <- cor(weighted_length_df$rbn, weighted_length_df$weighted_length, method = "spearman")
  cat("Gene weighted length and RBN correlation for sample", unique_sample, "(Spearman):\n", correlation, "\n")
}
```


